File: realsense/enhanced_photography/android/src/org/xwalk/extensions/realsense/enhancedphotography/PhotoUtilsObject.java
// Copyright (c) 2015 Intel Corporation. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.xwalk.extensions.realsense.enhancedphotography;

import android.util.Log;

import com.intel.camera.toolkit.depth.photography.core.DepthPhoto;
import com.intel.camera.toolkit.depth.photography.core.DPRect;
import com.intel.camera.toolkit.depth.photography.core.DPPoint;
import com.intel.camera.toolkit.depth.photography.utils.CommonFOV;
import com.intel.camera.toolkit.depth.photography.utils.Crop;
import com.intel.camera.toolkit.depth.photography.utils.DepthQuality;
import com.intel.camera.toolkit.depth.photography.utils.EnhanceDepth;
import com.intel.camera.toolkit.depth.photography.utils.EnhanceDepth.DepthEnhancementType;
import com.intel.camera.toolkit.depth.photography.utils.ResizeDepth;
import com.intel.camera.toolkit.depth.photography.utils.ResizePhoto;
import com.intel.camera.toolkit.depth.photography.utils.Rotate;

import java.lang.RuntimeException;
import java.util.UUID;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import org.xwalk.extensions.common.BindingObject;
import org.xwalk.extensions.common.BindingObjectStore;
import org.xwalk.extensions.common.FunctionInfo;

public class PhotoUtilsObject extends BindingObject {
    private static final String TAG = "PhotoUtilsObject";
    private BindingObjectStore mBindingObjectStore;

    public PhotoUtilsObject(BindingObjectStore bindingObjectStore) {
        mBindingObjectStore = bindingObjectStore;
        mHandler.register("colorResize", this);
        mHandler.register("commonFOV", this);
        mHandler.register("depthResize", this);
        mHandler.register("enhanceDepth", this);
        mHandler.register("getDepthQuality", this);
        mHandler.register("photoCrop", this);
        mHandler.register("photoRotate", this);
    }

    private void reportErrorMessage(String error, FunctionInfo info) {
        try {
            JSONArray result = new JSONArray();
            result.put(0, "");
            result.put(1, error);
            info.postResult(result);
        } catch (JSONException e) {
            Log.e(TAG, e.toString());
        }
    }

    private void createPhotoObjectAndReply(DepthPhoto photo, FunctionInfo info) {
        try {
            DepthPhotoObject photoObject = new DepthPhotoObject();
            photoObject.setDepthPhoto(photo);
            String objectId = UUID.randomUUID().toString();
            mBindingObjectStore.addBindingObject(objectId, photoObject);

            JSONArray result = new JSONArray();
            JSONObject photoJSONObject = new JSONObject();
            photoJSONObject.put("objectId", objectId);
            result.put(0, photoJSONObject);
            info.postResult(result);
        } catch (JSONException e) {
            Log.e(TAG, e.toString());
        }
    }

    public void onColorResize(FunctionInfo info) {
        try {
            JSONArray args = info.getArgs();
            JSONObject photo = args.getJSONObject(0);
            String photoId = photo.getString("objectId");
            DepthPhotoObject depthPhotoObject =
                    (DepthPhotoObject)mBindingObjectStore.getBindingObject(photoId);
            if (depthPhotoObject == null) {
                reportErrorMessage("Invalid DepthPhoto Object", info);
                return;
            }

            int width = args.getInt(1);
            DepthPhoto depthPhoto = depthPhotoObject.getDepthPhoto();
            try {
                DepthPhoto resizedPhoto = ResizePhoto.resizePhoto(depthPhoto, width);
                createPhotoObjectAndReply(resizedPhoto, info);
            } catch (RuntimeException e) {
                reportErrorMessage(e.toString(), info);
                return;
            }
        } catch (JSONException e) {
            Log.e(TAG, e.toString());
        }
    }

    public void onCommonFOV(FunctionInfo info) {
        try {
            JSONArray args = info.getArgs();
            JSONObject photo = args.getJSONObject(0);
            String photoId = photo.getString("objectId");
            DepthPhotoObject depthPhotoObject =
                    (DepthPhotoObject)mBindingObjectStore.getBindingObject(photoId);
            if (depthPhotoObject == null) {
                reportErrorMessage("Invalid DepthPhoto Object", info);
                return;
            }

            DepthPhoto depthPhoto = depthPhotoObject.getDepthPhoto();
            try {
                DepthPhoto fovPhoto = CommonFOV.getCroppedDepthPhoto(depthPhoto);
                createPhotoObjectAndReply(fovPhoto, info);
            } catch (RuntimeException e) {
                reportErrorMessage(e.toString(), info);
                return;
            }
        } catch (JSONException e) {
            Log.e(TAG, e.toString());
        }
    }

    public void onDepthResize(FunctionInfo info) {
        try {
            JSONArray args = info.getArgs();
            JSONObject photo = args.getJSONObject(0);
            String photoId = photo.getString("objectId");
            DepthPhotoObject depthPhotoObject =
                    (DepthPhotoObject)mBindingObjectStore.getBindingObject(photoId);
            if (depthPhotoObject == null) {
                reportErrorMessage("Invalid DepthPhoto Object", info);
                return;
            }

            int width = args.getInt(1);
            DepthPhoto depthPhoto = depthPhotoObject.getDepthPhoto();
            int originalWidth = (int)depthPhoto.getDepthMap().getDepthData().getWidth();
            int originalHeight = (int)depthPhoto.getDepthMap().getDepthData().getHeight();
            int height = (int)(width * originalHeight / originalWidth);
            try {
                DepthPhoto resizedPhoto = ResizeDepth.resizeDepth(depthPhoto, width, height);
                createPhotoObjectAndReply(resizedPhoto, info);
            } catch (RuntimeException e) {
                reportErrorMessage(e.toString(), info);
                return;
            }
        } catch (JSONException e) {
            Log.e(TAG, e.toString());
        }
    }

    public void onEnhanceDepth(FunctionInfo info) {
        try {
            JSONArray args = info.getArgs();
            JSONObject photo = args.getJSONObject(0);
            String photoId = photo.getString("objectId");
            DepthPhotoObject depthPhotoObject =
                    (DepthPhotoObject)mBindingObjectStore.getBindingObject(photoId);
            if (depthPhotoObject == null) {
                reportErrorMessage("Invalid DepthPhoto Object", info);
                return;
            }

            EnhanceDepth.DepthEnhancementType enhanceType =
                    EnhanceDepth.DepthEnhancementType.HIGH_QUALITY;
            String qulity = args.getString(1);
            if (qulity.equals("high")) {
                enhanceType = EnhanceDepth.DepthEnhancementType.HIGH_QUALITY;
            } else if (qulity.equals("low")) {
                enhanceType = EnhanceDepth.DepthEnhancementType.REAL_TIME;
            } else {
                reportErrorMessage("Invalid Depth Qulity", info);
                return;
            }
            DepthPhoto depthPhoto = depthPhotoObject.getDepthPhoto();
            try {
                DepthPhoto enhancedPhoto =
                        (new EnhanceDepth()).enhanceDepth(depthPhoto, enhanceType);
                createPhotoObjectAndReply(enhancedPhoto, info);
            } catch (RuntimeException e) {
                reportErrorMessage(e.toString(), info);
                return;
            }
        } catch (JSONException e) {
            Log.e(TAG, e.toString());
        }
    }

    public void onGetDepthQuality(FunctionInfo info) {
        try {
            JSONArray args = info.getArgs();
            JSONObject photo = args.getJSONObject(0);
            String photoId = photo.getString("objectId");
            DepthPhotoObject depthPhotoObject =
                    (DepthPhotoObject)mBindingObjectStore.getBindingObject(photoId);
            if (depthPhotoObject == null) {
                reportErrorMessage("Invalid DepthPhoto Object", info);
                return;
            }

            DepthPhoto depthPhoto = depthPhotoObject.getDepthPhoto();
            String replyQulity;
            try {
                DepthQuality.DepthQualityLevel qulity =
                        DepthQuality.evaluateDepthQuality(depthPhoto);
                if (qulity == DepthQuality.DepthQualityLevel.BAD) {
                    replyQulity = "bad";
                } else if (qulity == DepthQuality.DepthQualityLevel.FAIR) {
                    replyQulity = "fair";
                } else if (qulity == DepthQuality.DepthQualityLevel.GOOD) {
                    replyQulity = "good";
                } else {
                    reportErrorMessage("Unsupported depth quality", info);
                    return;
                }
            } catch (RuntimeException e) {
                reportErrorMessage(e.toString(), info);
                return;
            }

            JSONArray result = new JSONArray();
            result.put(0, replyQulity);
            info.postResult(result);
        } catch (JSONException e) {
            Log.e(TAG, e.toString());
        }
    }

    public void onPhotoCrop(FunctionInfo info) {
        try {
            JSONArray args = info.getArgs();
            JSONObject photo = args.getJSONObject(0);
            String photoId = photo.getString("objectId");
            DepthPhotoObject depthPhotoObject =
                    (DepthPhotoObject)mBindingObjectStore.getBindingObject(photoId);
            if (depthPhotoObject == null) {
                reportErrorMessage("Invalid DepthPhoto Object", info);
                return;
            }

            JSONObject rect = args.getJSONObject(1);
            int x = rect.getInt("x");
            int y = rect.getInt("y");
            int w = rect.getInt("w");
            int h = rect.getInt("h");
            DPRect dpRect = new DPRect();
            dpRect.setTopLeftPoint(new DPPoint(x, y));
            dpRect.setBottomRightPoint(new DPPoint(x + w, y + h));
            DepthPhoto depthPhoto = depthPhotoObject.getDepthPhoto();
            try {
                DepthPhoto cropPhoto = Crop.crop(depthPhoto, dpRect);
                createPhotoObjectAndReply(cropPhoto, info);
            } catch (RuntimeException e) {
                reportErrorMessage(e.toString(), info);
                return;
            }
        } catch (JSONException e) {
            Log.e(TAG, e.toString());
        }
    }

    public void onPhotoRotate(FunctionInfo info) {
        try {
            JSONArray args = info.getArgs();
            JSONObject photo = args.getJSONObject(0);
            String photoId = photo.getString("objectId");
            DepthPhotoObject depthPhotoObject =
                    (DepthPhotoObject)mBindingObjectStore.getBindingObject(photoId);
            if (depthPhotoObject == null) {
                reportErrorMessage("Invalid DepthPhoto Object", info);
                return;
            }

            DepthPhoto depthPhoto = depthPhotoObject.getDepthPhoto();
            try {
                // Currently, windows only support clock-wise 90-degree rotation
                // So, here we directly set the fixed parameters.
                // double rotation = args.getDouble("rotation");
                DepthPhoto rotatePhoto = Rotate.rotate(depthPhoto,
                                                       Rotate.RotateDirection.Clockwise,
                                                       Rotate.FixedDegrees.Degrees_90);
                createPhotoObjectAndReply(rotatePhoto, info);
            } catch (RuntimeException e) {
                reportErrorMessage(e.toString(), info);
                return;
            }
        } catch (JSONException e) {
            Log.e(TAG, e.toString());
        }
    }
}

File: realsense/enhanced_photography/win/photo_utils_object.cc
// Copyright (c) 2015 Intel Corporation. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "realsense/enhanced_photography/win/photo_utils_object.h"

#include <string>

#include "realsense/enhanced_photography/win/common_utils.h"
#include "realsense/enhanced_photography/win/depth_photo_object.h"

namespace realsense {
namespace enhanced_photography {

PhotoUtilsObject::PhotoUtilsObject(EnhancedPhotographyInstance* instance)
    : instance_(instance) {
  handler_.Register("colorResize",
                    base::Bind(&PhotoUtilsObject::OnColorResize,
                               base::Unretained(this)));
  handler_.Register("commonFOV",
                    base::Bind(&PhotoUtilsObject::OnCommonFOV,
                               base::Unretained(this)));
  handler_.Register("depthResize",
                    base::Bind(&PhotoUtilsObject::OnDepthResize,
                               base::Unretained(this)));
  handler_.Register("enhanceDepth",
                    base::Bind(&PhotoUtilsObject::OnEnhanceDepth,
                               base::Unretained(this)));
  handler_.Register("getDepthQuality",
                    base::Bind(&PhotoUtilsObject::OnGetDepthQuality,
                               base::Unretained(this)));
  handler_.Register("photoCrop",
                    base::Bind(&PhotoUtilsObject::OnPhotoCrop,
                               base::Unretained(this)));
  handler_.Register("photoRotate",
                    base::Bind(&PhotoUtilsObject::OnPhotoRotate,
                               base::Unretained(this)));

  session_ = PXCSession::CreateInstance();
  photo_utils_ = PXCEnhancedPhoto::PhotoUtils::CreateInstance(session_);
}

PhotoUtilsObject::~PhotoUtilsObject() {
  if (photo_utils_) {
    photo_utils_->Release();
    photo_utils_ = nullptr;
  }
  if (session_) {
    session_->Release();
    session_ = nullptr;
  }
}

void PhotoUtilsObject::OnColorResize(
    scoped_ptr<XWalkExtensionFunctionInfo> info) {
  jsapi::depth_photo::Photo photo;
  scoped_ptr<ColorResize::Params> params(
      ColorResize::Params::Create(*info->arguments()));
  if (!params) {
    info->PostResult(
        ColorResize::Results::Create(photo, "Malformed parameters"));
    return;
  }

  std::string object_id = params->photo.object_id;
  DepthPhotoObject* depthPhotoObject = static_cast<DepthPhotoObject*>(
      instance_->GetBindingObjectById(object_id));
  if (!depthPhotoObject || !depthPhotoObject->GetPhoto()) {
    info->PostResult(ColorResize::Results::Create(photo,
        "Invalid Photo object."));
    return;
  }

  DCHECK(photo_utils_);
  int width = params->width;
  PXCPhoto* pxcphoto =
      photo_utils_->ColorResize(depthPhotoObject->GetPhoto(), width);

  if (!pxcphoto) {
    info->PostResult(ColorResize::Results::Create(photo,
        "Failed to operate colorResize"));
    return;
  }

  CreateDepthPhotoObject(instance_, pxcphoto, &photo);
  info->PostResult(ColorResize::Results::Create(photo, std::string()));
}

void PhotoUtilsObject::OnCommonFOV(
    scoped_ptr<XWalkExtensionFunctionInfo> info) {
  jsapi::depth_photo::Photo photo;
  scoped_ptr<CommonFOV::Params> params(
      CommonFOV::Params::Create(*info->arguments()));
  if (!params) {
    info->PostResult(
        CommonFOV::Results::Create(photo, "Malformed parameters"));
    return;
  }

  std::string object_id = params->photo.object_id;
  DepthPhotoObject* depthPhotoObject = static_cast<DepthPhotoObject*>(
      instance_->GetBindingObjectById(object_id));
  if (!depthPhotoObject || !depthPhotoObject->GetPhoto()) {
    info->PostResult(CommonFOV::Results::Create(photo,
        "Invalid Photo object."));
    return;
  }

  DCHECK(photo_utils_);
  PXCPhoto* pxcphoto = photo_utils_->CommonFOV(depthPhotoObject->GetPhoto());

  if (!pxcphoto) {
    info->PostResult(CommonFOV::Results::Create(photo,
        "Failed to operate commonFOV"));
    return;
  }

  CreateDepthPhotoObject(instance_, pxcphoto, &photo);
  info->PostResult(CommonFOV::Results::Create(photo, std::string()));
}

void PhotoUtilsObject::OnDepthResize(
    scoped_ptr<XWalkExtensionFunctionInfo> info) {
  jsapi::depth_photo::Photo photo;
  scoped_ptr<DepthResize::Params> params(
      DepthResize::Params::Create(*info->arguments()));
  if (!params) {
    info->PostResult(
        DepthResize::Results::Create(photo, "Malformed parameters"));
    return;
  }

  std::string object_id = params->photo.object_id;
  DepthPhotoObject* depthPhotoObject = static_cast<DepthPhotoObject*>(
      instance_->GetBindingObjectById(object_id));
  if (!depthPhotoObject || !depthPhotoObject->GetPhoto()) {
    info->PostResult(DepthResize::Results::Create(photo,
        "Invalid Photo object."));
    return;
  }

  DCHECK(photo_utils_);
  int width = params->width;

  PXCPhoto* pxcphoto;
  if (params->quality) {
    PXCEnhancedPhoto::PhotoUtils::DepthFillQuality pxcquality;
    if (params->quality == DepthFillQuality::DEPTH_FILL_QUALITY_HIGH) {
      pxcquality = PXCEnhancedPhoto::PhotoUtils::DepthFillQuality::HIGH;
    } else {
      pxcquality = PXCEnhancedPhoto::PhotoUtils::DepthFillQuality::LOW;
    }
    pxcphoto = photo_utils_->DepthResize(depthPhotoObject->GetPhoto(),
                                         width,
                                         pxcquality);
  } else {
    pxcphoto = photo_utils_->DepthResize(depthPhotoObject->GetPhoto(), width);
  }

  if (!pxcphoto) {
    info->PostResult(DepthResize::Results::Create(photo,
        "Failed to operate DepthResize"));
    return;
  }

  CreateDepthPhotoObject(instance_, pxcphoto, &photo);
  info->PostResult(DepthResize::Results::Create(photo, std::string()));
}

void PhotoUtilsObject::OnEnhanceDepth(
  scoped_ptr<XWalkExtensionFunctionInfo> info) {
  jsapi::depth_photo::Photo photo;
  scoped_ptr<EnhanceDepth::Params> params(
      EnhanceDepth::Params::Create(*info->arguments()));
  if (!params) {
    info->PostResult(
        EnhanceDepth::Results::Create(photo, "Malformed parameters"));
    return;
  }

  std::string object_id = params->photo.object_id;
  DepthPhotoObject* depthPhotoObject = static_cast<DepthPhotoObject*>(
      instance_->GetBindingObjectById(object_id));
  if (!depthPhotoObject || !depthPhotoObject->GetPhoto()) {
    info->PostResult(EnhanceDepth::Results::Create(photo,
        "Invalid Photo object."));
    return;
  }

  DCHECK(photo_utils_);
  DepthFillQuality quality = params->quality;
  PXCEnhancedPhoto::PhotoUtils::DepthFillQuality pxcquality;
  if (quality == DepthFillQuality::DEPTH_FILL_QUALITY_HIGH)
    pxcquality = PXCEnhancedPhoto::PhotoUtils::DepthFillQuality::HIGH;
  else
    pxcquality = PXCEnhancedPhoto::PhotoUtils::DepthFillQuality::LOW;

  PXCPhoto* pxcphoto = photo_utils_->EnhanceDepth(depthPhotoObject->GetPhoto(),
                                                  pxcquality);
  if (!pxcphoto) {
    info->PostResult(EnhanceDepth::Results::Create(photo,
        "Failed to operate EnhanceDepth"));
    return;
  }

  CreateDepthPhotoObject(instance_, pxcphoto, &photo);
  info->PostResult(EnhanceDepth::Results::Create(photo, std::string()));
}

void PhotoUtilsObject::OnGetDepthQuality(
    scoped_ptr<XWalkExtensionFunctionInfo> info) {
  jsapi::photo_utils::DepthMapQuality depth_quality;
  scoped_ptr<GetDepthQuality::Params> params(
      GetDepthQuality::Params::Create(*info->arguments()));
  if (!params) {
    info->PostResult(GetDepthQuality::Results::Create(
        depth_quality, "Malformed parameters"));
    return;
  }

  std::string object_id = params->photo.object_id;
  DepthPhotoObject* depthPhotoObject = static_cast<DepthPhotoObject*>(
      instance_->GetBindingObjectById(object_id));
  if (!depthPhotoObject || !depthPhotoObject->GetPhoto()) {
    info->PostResult(GetDepthQuality::Results::Create(depth_quality,
        "Invalid Photo object."));
    return;
  }

  DCHECK(photo_utils_);
  PXCEnhancedPhoto::PhotoUtils::DepthMapQuality qulity =
      photo_utils_->GetDepthQuality(
          depthPhotoObject->GetPhoto()->QueryDepthImage());
  if (qulity == PXCEnhancedPhoto::PhotoUtils::DepthMapQuality::BAD) {
    depth_quality = DepthMapQuality::DEPTH_MAP_QUALITY_BAD;
  } else if (qulity == PXCEnhancedPhoto::PhotoUtils::DepthMapQuality::FAIR) {
    depth_quality = DepthMapQuality::DEPTH_MAP_QUALITY_FAIR;
  } else if (qulity == PXCEnhancedPhoto::PhotoUtils::DepthMapQuality::GOOD) {
    depth_quality = DepthMapQuality::DEPTH_MAP_QUALITY_GOOD;
  } else {
    info->PostResult(GetDepthQuality::Results::Create(depth_quality,
        "Unsupported depth quality."));
    return;
  }

  info->PostResult(GetDepthQuality::Results::Create(
      depth_quality, std::string()));
}

void PhotoUtilsObject::OnPhotoCrop(
    scoped_ptr<XWalkExtensionFunctionInfo> info) {
  jsapi::depth_photo::Photo photo;
  scoped_ptr<PhotoCrop::Params> params(
      PhotoCrop::Params::Create(*info->arguments()));
  if (!params) {
    info->PostResult(
        PhotoCrop::Results::Create(photo, "Malformed parameters"));
    return;
  }

  std::string object_id = params->photo.object_id;
  DepthPhotoObject* depthPhotoObject = static_cast<DepthPhotoObject*>(
      instance_->GetBindingObjectById(object_id));
  if (!depthPhotoObject || !depthPhotoObject->GetPhoto()) {
    info->PostResult(PhotoCrop::Results::Create(photo,
        "Invalid Photo object."));
    return;
  }

  DCHECK(photo_utils_);
  PXCRectI32 pxcrect;
  pxcrect.x = params->rect.x;
  pxcrect.y = params->rect.y;
  pxcrect.w = params->rect.w;
  pxcrect.h = params->rect.h;

  PXCPhoto* pxcphoto = photo_utils_->PhotoCrop(depthPhotoObject->GetPhoto(),
                                               pxcrect);
  if (!pxcphoto) {
    info->PostResult(PhotoCrop::Results::Create(photo,
        "Failed to operate photoCrop"));
    return;
  }

  CreateDepthPhotoObject(instance_, pxcphoto, &photo);
  info->PostResult(PhotoCrop::Results::Create(photo, std::string()));
}

void PhotoUtilsObject::OnPhotoRotate(
    scoped_ptr<XWalkExtensionFunctionInfo> info) {
  jsapi::depth_photo::Photo photo;
  scoped_ptr<PhotoRotate::Params> params(
      PhotoRotate::Params::Create(*info->arguments()));
  if (!params) {
    info->PostResult(
        PhotoRotate::Results::Create(photo, "Malformed parameters"));
    return;
  }

  std::string object_id = params->photo.object_id;
  DepthPhotoObject* depthPhotoObject = static_cast<DepthPhotoObject*>(
      instance_->GetBindingObjectById(object_id));
  if (!depthPhotoObject || !depthPhotoObject->GetPhoto()) {
    info->PostResult(PhotoRotate::Results::Create(photo,
        "Invalid Photo object."));
    return;
  }

  DCHECK(photo_utils_);

  pxcF32 rotation = params->rotation;
  PXCPhoto* pxcphoto = photo_utils_->PhotoRotate(depthPhotoObject->GetPhoto(),
                                                 rotation);
  if (!pxcphoto) {
    info->PostResult(PhotoRotate::Results::Create(photo,
        "Failed to operate photoRotate"));
    return;
  }

  CreateDepthPhotoObject(instance_, pxcphoto, &photo);
  info->PostResult(PhotoRotate::Results::Create(photo, std::string()));
}

}  // namespace enhanced_photography
}  // namespace realsense
